// @ledgerhq/types-live/lib-es/postOnboarding.js
/**
 * Unique identifier of a post onboarding action.
 */
export var PostOnboardingActionId;
(function (PostOnboardingActionId) {
    PostOnboardingActionId["deviceOnboarded"] = "deviceOnboarded";
    PostOnboardingActionId["assetsTransfer"] = "assetsTransfer";
    PostOnboardingActionId["buyCrypto"] = "buyCrypto";
    PostOnboardingActionId["syncAccounts"] = "syncAccounts";
    PostOnboardingActionId["customImage"] = "customImage";
    PostOnboardingActionId["recover"] = "recover";
    // Mocks for desktop development and tests
    PostOnboardingActionId["assetsTransferMock"] = "assetsTransferMock";
    PostOnboardingActionId["buyCryptoMock"] = "buyCryptoMock";
    PostOnboardingActionId["customImageMock"] = "customImageMock";
    PostOnboardingActionId["claimMock"] = "claimMock";
    PostOnboardingActionId["migrateAssetsMock"] = "migrateAssetsMock";
    PostOnboardingActionId["personalizeMock"] = "personalizeMock";
    PostOnboardingActionId["recoverMock"] = "recoverMock";
    PostOnboardingActionId["syncAccountsMock"] = "syncAccountsMock";
})(PostOnboardingActionId || (PostOnboardingActionId = {}));
