// @ledgerhq/client-ids/lib/api/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPushDevicesRequest = exports.pushDevicesApi = void 0;
const react_1 = require("@reduxjs/toolkit/query/react");
const live_env_1 = require("@ledgerhq/live-env");
/**
 * RTK Query API for Push Devices Service
 */
exports.pushDevicesApi = (0, react_1.createApi)({
    reducerPath: "pushDevicesApi",
    baseQuery: (0, react_1.retry)((0, react_1.fetchBaseQuery)({
        baseUrl: "",
        prepareHeaders: (headers) => {
            headers.set("Content-Type", "application/json");
            headers.set("X-Ledger-Client-Version", (0, live_env_1.getEnv)("LEDGER_CLIENT_VERSION"));
            return headers;
        },
    }), {
        maxRetries: 3,
    }),
    tagTypes: [],
    endpoints: build => ({
        /**
         * Push devices to the backend service
         * This endpoint updates the backend with the current list of device IDs for a user
         * Success is determined by HTTP response code (2xx), no response body is expected
         */
        pushDevices: build.mutation({
            query: body => {
                const baseUrl = (0, live_env_1.getEnv)("PUSH_DEVICES_SERVICE_URL");
                return {
                    url: `${baseUrl}/pushdevices`,
                    method: "POST",
                    body,
                };
            },
        }),
    }),
});
/**
 * Create a push devices request
 * @param userId - User ID (equipment_id) as string (managed by apps)
 * @param deviceIds - Array of DeviceId objects
 */
function createPushDevicesRequest(userId, deviceIds) {
    const equipment_id = userId;
    const devices = deviceIds.map(deviceId => deviceId.exportDeviceIdForPushDevicesService());
    return {
        equipment_id,
        devices,
    };
}
exports.createPushDevicesRequest = createPushDevicesRequest;
